### PR TITLE
Only build if new "build-flac"/"build-ogg" features are selected

### PR DESCRIPTION
**Pull request by nabijaczleweli**

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -31,6 +31,29 @@ jobs:
...
+          sudo apt-get install --yes --no-install-recommends libflac-dev libogg-dev
+      - name: Clone Git repository
+        uses: actions/checkout@v2
+        with:
+          submodules: true
+      - name: Install Rust
+        uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          toolchain: stable${{ matrix.toolchain-suffix }}
```
Comment (mgeier) [inline .github/workflows/main.yml:51]: The `matrix` stuff is not relevant here.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -31,6 +31,29 @@ jobs:
         run: |
           cargo test --workspace --all-features
       - name: Run tests (without OGG)
+        run: |
+          cargo test --workspace --no-default-features --features="build"
+
+  tests-distribution-libflac:
```
Comment (mgeier) [inline .github/workflows/main.yml:37]: It took me a while to understand what "distribution" means in this case.
I was first thinking that something is distributed here, which it's of course not.

Maybe rename it to something with "external"?
Or is there a better word?

File: build.rs [hunk 3]
```diff
@@ -1,3 +1,12 @@
+#[cfg(not(feature = "build"))]
+fn main() {
+    if cfg!(feature = "ogg") {
+        println!("cargo:rustc-link-lib=ogg");
```
Comment (mgeier) [inline build.rs:4]: Is this really necessary?
It somehow contradicts the change in `tests/ogg_support.rs` doesn't it?

File: tests/ogg_support.rs [hunk 4]
```diff
@@ -1,10 +1,9 @@
-use libflac_sys as ffi;
-
 #[test]
+#[cfg(feature = "build")]
```
Comment (mgeier) [inline tests/ogg_support.rs:2]: I was wondering why this is necessary, but I tried it and could confirm that this doesn't work when `build` is off, `ogg` is off but the external `libFLAC` indeed *does* support OGG.

So I guess this line here is necessary, but is linking to `libogg` necessary in `build.rs`?

File: src/lib.rs [hunk 5]
```diff
@@ -22,11 +22,14 @@ libflac-sys = "0.1"
 # Features
 
 * `ogg` (enabled by default): Support for FLAC data in OGG containers
```
Comment (mgeier) [inline src/lib.rs:24]: Maybe we should add that this is ignored if `build` is off?

File: src/lib.rs [hunk 6]
```diff
@@ -22,11 +22,14 @@ libflac-sys = "0.1"
...
+* `build` (enabled by default): build instead of linking to libraries from
+   the system
```
Comment (mgeier) [inline src/lib.rs:26]: What happens if the external `libFLAC` is incompatible with the headers used for creating the Rust bindings?

Should we add a warning here?

File: .github/workflows/main.yml [hunk 7]
```diff
@@ -31,6 +31,29 @@ jobs:
...
+          submodules: true
+      - name: Install Rust
+        uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          toolchain: stable${{ matrix.toolchain-suffix }}
+          override: true
+      - name: Run tests against distribution libraries
+        run: |
+          cargo test --workspace --no-default-features --features="ogg"
```
Comment (mgeier) [inline .github/workflows/main.yml:55]: I'm not sure whether the `ogg` feature makes a difference here?

File: src/lib.rs [hunk 8]
```diff
@@ -22,11 +22,14 @@ libflac-sys = "0.1"
...
 
 # Building the `libFLAC` and `libogg` libraries
 
-When building this crate, the `libFLAC` library is automatically built as well,
-using the [cmake] crate.
+When building this crate with the `build` feature,
+the `libFLAC` library is automatically built as well,
+using the [cmake] crate – `cmake` and a C toolchain is required
```
Comment (mgeier) [inline src/lib.rs:32]: Missing dot at the end.

**Review by mgeier — COMMENTED** [review]
Thanks a lot, that's great!

File: build.rs (hunk 3 above)
Comment (mgeier) [reply build.rs:4]: Or is this necessary for static linking (I tried only dynamic linking)?

File: .github/workflows/main.yml (hunk 2 above)
Comment (nabijaczleweli) [reply .github/workflows/main.yml:37]: It was a coin-flip for me between "distribution" and "external"; "external" it is

File: src/lib.rs (hunk 6 above)
Comment (nabijaczleweli) [reply src/lib.rs:26]: I don't think this is a concern – going by https://packages.debian.org/search?keywords=flac, jessie is on 1.3.0, stretch and buster are on 1.3.2, and sid is on 1.3.3, all of which are compatible. Also https://xiph.org/flac/ has a news sexion, which has these dates on it:
```
4 Aug 2019 : FLAC 1.3.3 released
1 Jan 2017 : FLAC 1.3.2 released
27 Nov 2014: FLAC 1.3.1 released
01 Jun 2013: FLAC 1.3.0 released
```

The likelihood of there being a new incompatible version is too close to zero IMO.

File: tests/ogg_support.rs (hunk 4 above)
Comment (nabijaczleweli) [reply tests/ogg_support.rs:2]: It's not, I appear to have been confused initially so as to how libogg factored into all this.

File: .github/workflows/main.yml (hunk 7 above)
Comment (nabijaczleweli) [reply .github/workflows/main.yml:55]: It doesn't, and we promise it doesn't. Let's make sure we keep that promise.

**Comment by nabijaczleweli** [comment]
Updated

File: build.rs (hunk 3 above)
Comment (mgeier) [reply build.rs]: To answer my own question:

> Or is this necessary for static linking?

No. I just tried it by disabling my local `libFLAC.so`, leaving only `libFLAC.a`, which still gave me OGG support.

File: tests/ogg_support.rs (hunk 4 above)
Comment (mgeier) [reply tests/ogg_support.rs:2]: OK, thanks, now the contradiction is resolved.

File: src/lib.rs (hunk 6 above)
Comment (mgeier) [reply src/lib.rs:25]: OK, thanks, makes sense.

File: .github/workflows/main.yml (hunk 7 above)
Comment (mgeier) [reply .github/workflows/main.yml:55]: But I don't understand what we are testing here.

The OGG test is disabled in this case, isn't it?

**Comment by mgeier** [comment]
I just had an idea: what if we rename the features to `build-flac` and `build-ogg`?

I think this would be clearer, and it would avoid the confusion of not being able to control the OGG support of an external `libFLAC`.

We could do something like:

```toml
[features]
default = ["build-flac", "build-ogg"]
build-flac = ["cmake"]
build-ogg = ["build-flac"]
```

It would also encode explicitly in the `Cargo.toml` that building OGG is impossible without building FLAC.

What do you think?

File: .github/workflows/main.yml (hunk 7 above)
Comment (nabijaczleweli) [reply .github/workflows/main.yml:55]: That `ogg` really is ignored and doesn't try to link to libogg, which is not installed.

**Comment by nabijaczleweli** [comment]
That does indeed make more sense. Updated

File: src/lib.rs [hunk 9]
```diff
@@ -21,16 +21,12 @@ libflac-sys = "0.1"
...
-using the [cmake] crate.
-
-[cmake]: https://crates.io/crates/cmake
-
-Furthermore, the `libogg` library is built when the `ogg` feature is selected.
+* `build-flac` (enabled by default): build `libFLAC` instead of linking
+   to the system library – `cmake` and a C toolchain is required
+* `build-ogg` (enabled by default, implies `build-flac`):
+   build support for FLAC data in OGG containers into `libFLAC`;
+   if the `build-flac` is not selected, support for OGG containers
```
Comment (mgeier) [inline src/lib.rs:28]: The `the` is superfluous, right?

**Review by mgeier — COMMENTED** [review]
Thanks a lot, this looks even better now.

There is just a last very little nit to pick ...

File: .github/workflows/main.yml (hunk 7 above)
Comment (mgeier) [reply .github/workflows/main.yml]: This problem doesn't exist anymore.

File: src/lib.rs (hunk 9 above)
Comment (nabijaczleweli) [reply src/lib.rs:28]: That was gonna say "the `build-flac` feature" but just "`build-flac`" is better, yes

**Comment by nabijaczleweli** [comment]
Updated

File: .github/workflows/main.yml (hunk 7 above)
Comment (nabijaczleweli) [reply .github/workflows/main.yml]: Which is why that line is no longer

**Comment by mgeier** [comment]
Perfect, thanks a lot for your work!

File: .github/workflows/main.yml (hunk 7 above)
Comment (mgeier) [reply .github/workflows/main.yml]: I know, I just wanted to acknowledge that, in case somebody else reads that in the future ...

**Pull request merged**